### DISCUSSION TITLE
How do I make a quiz?

### DISCUSSION THREAD

**Discussion by smashah** [open-wa/wa-automate-nodejs#2779] [Q&A]
Lets say you have 3 questions, `age range`, `gender` and `country`:

You can encode the button ID with the quiz state relating to that answer. So let's say we start the quiz by asking the age range with 3 standard quick reply buttons. The buttons will be as follows:

| Button # | Button Value | Quiz state | Button ID (Base64 encoded quiz state) |
|----|----|----|----|
|1.| 0-15 | {"age_range":"0-15"} | `eyJhZ2VfcmFuZ2UiOiIwLTE1In0=` |
|2.| 16-22 |{"age_range":"16-22"} | `eyJhZ2VfcmFuZ2UiOiIxNi0yMiJ9` |
|3.| 23-40 | {"age_range":"23-40"} | `eyJhZ2VfcmFuZ2UiOiIyMy00MCJ9` |

As you can see, the base64 encoded button ID is the end state after answering the question with the relevant answer.

The code will be as follows:

```javascript
await client.sendButtons(chatId, "What age are you?", [
"id": "eyJhZ2VfcmFuZ2UiOiIwLTE1In0=",
"text" : "0-15"
],[
"id": "eyJhZ2VfcmFuZ2UiOiIxNi0yMiJ9",
"text" : "16-22"
],[
"id": "eyJhZ2VfcmFuZ2UiOiIyMy00MCJ9",
"text" : "23-40"
], "1/3")
```

Let's assume the user selected the first option `"0-15"`, the next question is `gender` with three more options.

| Button # | Button Value | Quiz state | Button ID (Base64 encoded quiz state) |
|----|----|----|----|
|1.| male | {"age_range":"0-15", "gender": "male"} | `eyJhZ2VfcmFuZ2UiOiIwLTE1IiwgImdlbmRlciI6ICJtYWxlIn0=` |
|2.| female |{"age_range":"0-15", "gender": "female"} | `eyJhZ2VfcmFuZ2UiOiIwLTE1IiwgImdlbmRlciI6ICJmZW1hbGUifQ==` |
|3.| other | {"age_range":"0-15", "gender": "other"} | `eyJhZ2VfcmFuZ2UiOiIwLTE1IiwgImdlbmRlciI6ICJvdGhlciJ9` |

```javascript
await client.sendButtons(chatId, "Select gender", [
"id": "eyJhZ2VfcmFuZ2UiOiIwLTE1IiwgImdlbmRlciI6ICJtYWxlIn0=",
"text" : "male"
],[
"id": "eyJhZ2VfcmFuZ2UiOiIwLTE1IiwgImdlbmRlciI6ICJmZW1hbGUifQ==",
"text" : "female"
],[
"id": "eyJhZ2VfcmFuZ2UiOiIwLTE1IiwgImdlbmRlciI6ICJvdGhlciJ9",
"text" : "other"
], "2/3")
```

The base64 buttons IDs are naturally longer because the quizstate object has more data. Assuming the user selected 1, the button ID of the response will be `eyJhZ2VfcmFuZ2UiOiIwLTE1IiwgImdlbmRlciI6ICJtYWxlIn0=`.

Now, onto the last question, `country`. There are 195 countries in the world. Let's assume our users are in one of 20 countries. You cannot use `sendButtons` with 20 options. Luckily, you can use [`sendListMessage`](https://docs.openwa.dev/classes/api_Client.Client.html#sendListMessage) instead.

Now we can introduce some quality of life improvements to dynamically generate the required list message.

```javascript
//this function converts an object to a base 64 string
const b64ify = (obj) => Buffer.from(JSON.stringify(obj)).toString('base64');
const countries = [
	"Afghanistan",
	"Albania",
	"Algeria",
	"American Samoa",
	"Andorra",
	"Angola",
	"Anguilla",
	"Antarctica",
	"Antigua and Barbuda",
	"Argentina",
	"Armenia",
	"Aruba",
	"Australia",
	"Austria",
	"Azerbaijan",
	"Bahamas (the)",
	"Bahrain",
	"Bangladesh",
	"Barbados"
]
```

```javascript
await client.sendListMessage(chatId, [
    title: "Countries",
    rows: countries.map(country => ({
        title: country,
        rowId: b64ify({
            ...currentState,
            country
        })
    }))
], "Please select a country", "3/3", "Choose")
```

And voila, when you get the response for this last list message, you can determine that the quiz has ended (because all properties are answered) and you've done it without integrating a database to save conversation/quiz state and without integrating dialogflow (which is not free at scale nor stateful beyond 5 minutes).

## What's next?

Ideally, there should be an inbuilt quiz management which dynamically does the background work for you and uses a message collector type system to emit finalized quizzes for you.

[TBC]

Ref (from discord discussion):

https://github.com/open-wa/wa-automate-nodejs/issues/1794#issuecomment-955841339

**Comment by swatky** [comment]
Very nice, this can be used for many things.
I think it's possible to make a "Feedback" of user satisfaction with the Bot, the range of possibilities is very wide, in my opinion.
